lib/hanami/db/testing.rb

```diff
@@ -16,18 +17,128 @@ module Testing
       DATABASE_NAME_MATCHER = /_dev(elopment)?$/
       private_constant :DATABASE_NAME_MATCHER
 
-      # @api private
-      # @since 2.2.0
-      def self.database_url(url)
-        url = URI(url.to_s)
+      class << self
+        # @api private
+        # @since 2.2.0
+        def database_url(url)
+          url = parse_url(url)
+
+          case deconstruct_url(url)
+          in { scheme: "sqlite", opaque: nil, path: } unless path.nil?
+            url.path = database_filename(path)
+          in { path: String => path } if path =~ DATABASE_NAME_MATCHER
+            url.path = path.sub(DATABASE_NAME_MATCHER, DATABASE_NAME_SUFFIX)
+          in { path: String => path } unless path.end_with?(DATABASE_NAME_SUFFIX)
+            url.path << DATABASE_NAME_SUFFIX
+          else
+            # do nothing
+          end
+
+          stringify_url(url)
+        end
+
+        private
+
+        # @api private
+        # @since 2.2.0
+        def parse_url(url)
+          if url.is_a?(URI::Generic)
+            # URI#dup does not duplicate internal instance
+            # variables, making mutation dangerous.
+            URI(stringify_url(url))
+          else
+            URI(url.to_s)
+          end
+        end
+
+        # rubocop:disable Metrics/AbcSize, Metrics/PerceivedComplexity
+
+        # Work around a bug in Ruby 3.0.x that erroneously omits
+        # the '//' prefix from hierarchical URLs.
+        #
+        # @api private
+        # @since 2.2.0
+        def stringify_url(url)
+          if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("3.1.0")
+            return url.to_s
+          end
+
+          require "stringio"
+
+          buf = StringIO.new
```

Comment:
Curious why you're using a StringIO here instead of a plain old String?

Reply:
Better performance for lots of mutations, just something I do by default in cases like this but it's _probably_ fine either way.